crates/uv/src/commands/tool/common.rs

```diff
@@ -169,129 +174,147 @@ pub(crate) async fn refine_interpreter(
         executable_directory.user_display()
     );
 
-    let entry_points = entrypoint_paths(
-        &site_packages,
-        installed_dist.name(),
-        installed_dist.version(),
-    )?;
-
-    // Determine the entry points targets. Use a sorted collection for deterministic output.
-    let target_entry_points = entry_points
+    let mut installed_entrypoints = Vec::new();
+    let site_packages = SitePackages::from_environment(environment)?;
+    let ordered_packages = entrypoints
+        // Install dependencies first
         .into_iter()
-        .map(|(name, source_path)| {
-            let target_path = executable_directory.join(
-                source_path
-                    .file_name()
-                    .map(std::borrow::ToOwned::to_owned)
-                    .unwrap_or_else(|| OsString::from(name.clone())),
-            );
-            (name, source_path, target_path)
-        })
-        .collect::<BTreeSet<_>>();
+        .filter(|pkg| pkg != name)
+        .collect::<BTreeSet<_>>()
+        // Then install the main package last
+        .into_iter()
+        .chain(iter::once(name.clone()));
 
-    if target_entry_points.is_empty() {
-        writeln!(
-            printer.stdout(),
-            "No executables are provided by package `{from}`; removing tool",
-            from = name.cyan()
-        )?;
+    for package in ordered_packages {
+        if package == *name {
+            debug!("Installing entrypoints for tool `{package}`");
+        } else {
+            debug!("Installing entrypoints for `{package}` as part of tool `{name}`");
+        }
 
-        hint_executable_from_dependency(name, &site_packages, printer)?;
+        let installed = site_packages.get_packages(&package);
+        let dist = installed
+            .first()
+            .context("Expected at least one requirement")?;
+        let dist_entrypoints = entrypoint_paths(&site_packages, dist.name(), dist.version())?;
 
-        // Clean up the environment we just created.
-        installed_tools.remove_environment(name)?;
+        // Determine the entry points targets. Use a sorted collection for deterministic output.
+        let target_entrypoints = dist_entrypoints
+            .into_iter()
+            .map(|(name, source_path)| {
+                let target_path = executable_directory.join(
+                    source_path
+                        .file_name()
+                        .map(std::borrow::ToOwned::to_owned)
+                        .unwrap_or_else(|| OsString::from(name.clone())),
+                );
+                (name, source_path, target_path)
+            })
+            .collect::<BTreeSet<_>>();
 
-        return Ok(ExitStatus::Failure);
-    }
+        if target_entrypoints.is_empty() {
+            writeln!(
+                printer.stdout(),
+                "No executables are provided by package `{}`; removing tool `{name}`",
+                package.cyan()
+            )?;
```

Comment:
Does this mean we'll fail if _any_ of the `--with-executables-from` packages doesn't have entry points? It should probably just be a warning hinting to use `--with` instead if it's not the root package.